cmd/tailscaled/tailscaled.go

```diff
@@ -432,13 +432,25 @@ func startIPNServer(ctx context.Context, logf logger.Logf, logID logid.PublicID,
 	if sigPipe != nil {
 		signal.Ignore(sigPipe)
 	}
+	wgEngineCreated := make(chan struct{})
 	go func() {
-		select {
-		case s := <-interrupt:
-			logf("tailscaled got signal %v; shutting down", s)
-			cancel()
-		case <-ctx.Done():
-			// continue
+		var wgEngineClosed <-chan struct{}
+		for {
+			select {
+			case s := <-interrupt:
+				logf("tailscaled got signal %v; shutting down", s)
+				cancel()
+				return
+			case <-wgEngineClosed:
+				logf("wgengine has been closed; shutting down")
+				cancel()
+				return
+			case <-wgEngineCreated:
+				wgEngineClosed = sys.Engine.Get().Wait()
+				wgEngineCreated = nil
```

Comment:
this line's slightly shady data-race-wise. It was created and closed by the other goroutine so zeroing it out here (for the local effect of it not being readable anymore) would normally be bad except for the fact that it's actually safe because nobody else currently reads from it and it only has one writer (the one close).

But if something else later went to read from it, it'd be a race.

Safe fix is to add at the top of this goroutine (at L438):

```
   wgEngineCreated := wgEngineCreated // local shadow
```

Reply:
Good point! Thank you and fixed.